### PR TITLE
#17 - Invoke truststore.populate outside of the surrounding if

### PR DESCRIPTION
**Pull request by RomanKisilenko**
Please review this PR.

### CODE REVIEW

**Comment by dmarkov** [comment]
Thanks for the pull request, let me find a reviewer..

**Comment by dmarkov** [comment]
@longtimeago this pull request is for you, please review

**Comment by longtimeago** [comment]
@RomanKisilenko thank you

**Comment by longtimeago** [comment]
@rultor good to merge

**Comment by yegor256** [comment]
@RomanKisilenko we're fixing something here, but why didn't we reproduce it first in a test?

**Comment by RomanKisilenko** [comment]
@yegor256 The implementation is not testable. E.g. I can not mock Keystore and Cacerts to do behavioral testing due to classes are final. What should I do in situations like this?

**Comment by yegor256** [comment]
@RomanKisilenko do you really need to mock them? Why not just use normal instances of these classes?

**Comment by RomanKisilenko** [comment]
@yegor256 Mocking makes the test much less complicated. This is just a unit test, so we should not verify how Keystore and Cacerts work. Instead we should use test doubles to verify interaction. Should I proceed with implementing the test using real Keystore and Cacerts instances?

**Comment by yegor256** [comment]
@RomanKisilenko mocking makes tests much more complicated, see this article about this subject: http://www.yegor256.com/2014/09/23/built-in-fake-objects.html So, yes, it's better use real instances of `Keystore` and `Cacerts`

**Comment by RomanKisilenko** [comment]
@yegor256 I attempted implementing unit test using the live Keystore and Cacerts and run into following problem, due to objects are too much coupled and thus not testable.
Keytool.createLocaleDependentYes throws java.util.MissingResourceException due to bundle used not found. I have not found a way to work around this issue.
And overall test looks a bit more complicated in this case (the number of lines have grown compared to interaction-based test), so I went further and implemented standard interaction-based test by making Keystore and Cacerts classes non final. Hope it makes sense for you.

**Comment by yegor256** [comment]
@RomanKisilenko I didn't get the idea behind this factory (which is a terrible name for a class, btw). Why can't you just pass `keystore` and `cacerts` to the ctor of the mojo?

**Comment by RomanKisilenko** [comment]
@yegor256 It looks like making classes not final is not possible, maven throws an error. Reimplemented the test.

File: src/main/java/com/jcabi/ssl/maven/plugin/KeygenMojo.java [hunk 1]
```diff
@@ -110,21 +142,25 @@ public void execute() throws MojoFailureException {
...
-        if (!store.isActive()) {
-            try {
-                store.activate(this.keystore);
-                final Cacerts truststore = new Cacerts(this.cacerts);
-                truststore.imprt();
-                truststore.populate(this.project.getProperties());
-            } catch (final java.io.IOException ex) {
-                throw new IllegalStateException(ex);
+        if (this.store == null) {
+            this.store = new Keystore(
```
Comment (yegor256) [inline src/main/java/com/jcabi/ssl/maven/plugin/KeygenMojo.java]: this should happen in ctor

File: src/main/java/com/jcabi/ssl/maven/plugin/KeygenMojo.java [hunk 2]
```diff
@@ -110,21 +142,25 @@ public void execute() throws MojoFailureException {
...
+                DigestUtils.md5Hex(KeygenMojo.class.getName())
+            );
+        }
+        try {
+            if (this.truststore == null) {
+                this.truststore = new Cacerts(this.cacerts);
```
Comment (yegor256) [inline src/main/java/com/jcabi/ssl/maven/plugin/KeygenMojo.java]: this too

**Comment by yegor256** [comment]
@RomanKisilenko null is evil, don't use it ever, see http://www.yegor256.com/2014/05/13/why-null-is-bad.html

**Comment by RomanKisilenko** [comment]
@yegor256 The store and truststore object are constructed based on injected fields, inside constructor they are null, so we can not construct store and truststore inside constructor.

**Comment by yegor256** [comment]
@RomanKisilenko `KeygenMojo()` should call another ctor with parameters, which will be set to default values. see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#constructors

**Comment by RomanKisilenko** [comment]
@yegor256 I mean we can not construc truststore in constructor because cacerts field value is not set at this point, so this is not about how constructor is implemented, this is about how maven injects field values.

**Comment by yegor256** [comment]
@RomanKisilenko Maven calls default ctor and it will make an instance of `Keystore` and call 3-args ctor.

**Comment by RomanKisilenko** [comment]
@yegor256 Fixed, please check.

**Comment by yegor256** [comment]
@rultor merge pls

**Comment by dmarkov** [comment]
@longtimeago Many thanks! **17 mins** were added to your account in Transaction ID `50907610`

+17 to your rating, your total score is [+1462](http://www.netbout.com/b/35024?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy now pls

**Pull request merged**